### PR TITLE
Command-system pr

### PR DESCRIPTION
**Pull request by Tais993**
This will be the main PR for the command-system.
See the relating issue #10 for more information.

There's still a few things we have to do, therefor:

- [X] Add handling for buttons and selection menus in the CommandHandler class
- [x] Util for generating component ID's of commands
- [x] Util for converting component ID's of commands to a list - will be pushed when internet is available
- [x] Add command which updates slash-commands
- [x] Command handler 

more in the future..

### CODE REVIEW

**Comment by Tais993** [comment]
While I originally wanted to add functionality so it's easier to work with slash-command permissions.
I myself will wait for the rework of the slash-command permissions.

Current way slash-command permissions work is shit, and Discord has confirmed on reworking them.
See [this discussion](https://github.com/discord/discord-api-docs/discussions/3581), the moderation section 
"Revisit the permissions system to make it more flexible"

So anyone is free to create it, but most likely you will have to revamp it when the reworked permission system comes.

**Review by illuminator3 — CHANGES_REQUESTED** [review]
Please apply _our_ formatting by running `gradle spotlessApply` :)

File: application/src/main/java/org/togetherjava/tjbot/commands/CommandHandler.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public void onButtonClick(@NotNull ButtonClickEvent event) {
+        String[] argsArray = event.getComponentId().split("-");
+        Command command = commandMap.get(argsArray[0]);
+
+        if (command != null) {
+            List<String> args = new ArrayList<>(Arrays.asList(argsArray));
+            args.remove(0);
+            args.remove(0);
+            command.onButtonClick(event, args);
```
[the remark this answers is not in the document]
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/CommandHandler.java:60]: Didn't think about that, and that's why reviews are always appreciated :p

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                    event.reply("Next time use the right command!").queue();
+                    event.getMessage().editMessageComponents().queue();
+                }
+
+                case SUCCESS -> {
+                    event.deferReply().queue();
+                    List<Command> commands = commandHandler.getCommandList();
+
+                    CommandListUpdateAction globalCommandsUpdateAction = event.getJDA().updateCommands();
+                    CommandListUpdateAction guildCommandsUpdateAction = event.getJDA().getGuilds().get(0).updateCommands();
```
[the remark this answers is not in the document]
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java:76]: I should just loop through the guilds and add commands, always works.

[the remark this answers is not in the document]
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java]: Updated it, code looks less readable but it works in every guild now.

**Review by illuminator3 — APPROVED** [review]

**Comment by Tais993** [comment]
I'll accept after 2 reviews, not the required 2 since I find this really important.

**Comment by Tais993** [comment]
> Apart from the added review comments. I think extracting out some of very large lambdas into well named functions will aid in readability as well as testability.

You'll have to quote some exact lambdas, I find it myself readable and maintainable this way.

File: application/src/main/java/org/togetherjava/tjbot/commands/CommandHandler.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        Command command = commandMap.get(event.getName());
+
+        if (command != null) {
+            command.onSlashCommand(event);
+        } else {
+            logger.error("Command '{}' doesn't exist?", event.getName());
+        }
+    }
+
+    @Override
```
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/CommandHandler.java:48]: You're repeating yourself [here](https://github.com/Together-Java/TJ-Bot/pull/37/files#diff-e5f33a13deeba755c63c3da6a06821b5b943b870e8a92c7b05ed588da51ba6dbR48-R70), I'd say extract this to a separate method.

File: .github/workflows/docker-publish.yaml [hunk 4]
```diff
@@ -3,17 +3,19 @@ name: Docker Publish
 on:
   push:
     branches:
-      - 'master'
```
Comment (d3vel0per) [inline .github/workflows/docker-publish.yaml:6]: Why did we remove master or are we saying we will just use images from develop instead ie. for prod deployments?

Comment (I-Al-Istannen) [reply .github/workflows/docker-publish.yaml:6]: That's not part of this PR but incidental due to the git structure of it. I removed it as we need two different image tags (to ensure we can run both develop and prod in parallel) and therefore need to overhaul that anyways.

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                }
+
+                case SUCCESS -> {
+                    event.deferReply().queue();
+                    List<Command> commands = commandHandler.getCommandList();
+                    List<RestAction<List<net.dv8tion.jda.api.interactions.commands.Command>>> restActions =
+                            new ArrayList<>();
+
+                    // * adds all the global commands
+                    CommandListUpdateAction globalCommandsUpdateAction =
```
Comment (d3vel0per) [inline application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java:77]: This is with regards commentary around extracting methods from large lambdas eg.
```java
private final RestAction<List<net.dv8tion.jda.api.interactions.commands.Command>> restActionsForGlobalCommandUpdate(Event, List<Command>)
```

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                            globalCommandsUpdateAction.addCommands(
+                                    command.addOptions(new CommandData(command.getCommandName(),
+                                            command.getDescription())));
+                        }
+                    });
+
+                    restActions.add(globalCommandsUpdateAction);
+
+                    // * adds all the guild commands to each guild
+                    event.getJDA().getGuildCache().forEach(guild -> {
```
Comment (d3vel0per) [inline application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java:91]: Similar to the previous comment a similar one for guild specific commands

File: application/src/main/java/org/togetherjava/tjbot/commands/CommandHandler.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        getArgumentsAndCommandByEvent(event, (command, args) -> command.onButtonClick(event, args));
+    }
+
+    @Override
+    public void onSelectionMenu(@NotNull SelectionMenuEvent event) {
+        getArgumentsAndCommandByEvent(event,
+                (command, args) -> command.onSelectionMenu(event, args));
+    }
+
+    public void getArgumentsAndCommandByEvent(GenericComponentInteractionCreateEvent event,
```
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/CommandHandler.java:79]: this is missing javadoc

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                            .editOriginal(
+                                    "Commands successfully reloaded! *Global commands can take upto 1 hour to load")
+                            .queue();
+                    });
+                }
+            }
+        }
+    }
+
+    public CommandListUpdateAction getGlobalCommandUpdateRestAction(Event event,
```
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java:89]: same as above

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            List<Command> commands) {
+        return addCommandsToUpdateAction(event.getJDA().updateCommands(), commands, false);
+    }
+
+    public List<CommandListUpdateAction> getGuildCommandUpdateRestActions(Event event,
```
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java:94]: same as above

**Review by illuminator3 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java (hunk 8 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java]: While the methods were public, they aren't for use, it's a command not something public.
So I don't see any points in making docs, the code itself is extremely simple,
It feels like creating docs for methods like this is just a waste of everybody's time

File: application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java (hunk 9 above)
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/ReloadCommand.java]: See comment above

**Comment by Tais993** [comment]
2 things need to be fixed atm

Naming of the AvstractCommand is incorrect and bad, alternatives?

Docs mention using the interface still

**Review by I-Al-Istannen — APPROVED** [review]
Looks good for now, we'll see where we might need to take it :)

**Review by marko-radosavljevic — APPROVED** [review]

**Pull request merged**